src/meltano/cli/utils.py

```diff
@@ -528,44 +530,141 @@ def check_dependencies_met(
 
 
-class InstrumentedDefaultGroup(DefaultGroup):
-    """A variation of a DefaultGroup that instruments its invocation by updating the telemetry context."""
+def activate_environment(
+    ctx: click.Context, project: Project, required: bool = False
+) -> None:
+    """Activate the selected environment.
 
-    def invoke(self, ctx):
-        """Update the telemetry context and invoke the group."""
-        ctx.ensure_object(dict)
-        if ctx.obj.get("tracker"):
-            ctx.obj["tracker"].add_contexts(CliContext.from_click_context(ctx))
-        super().invoke(ctx)  # noqa: WPS608
+    The selected environment is whatever was selected with the `--environment`
+    option, or the default environment (set in `meltano.yml`) otherwise.
 
+    Args:
+        ctx: The Click context, used to determine the selected environment.
+        project: The project for which the environment will be activated.
+    """
+    if ctx.obj["selected_environment"]:
+        project.activate_environment(ctx.obj["selected_environment"])
+        # Update the project context being used for telemetry:
+        project_ctx = next(
+            ctx
+            for ctx in ctx.obj["tracker"].contexts
+            if isinstance(ctx, ProjectContext)
+        )
+        project_ctx.environment_name = ctx.obj["selected_environment"]
 
-class InstrumentedGroup(click.Group):
-    """A click.Group that instruments its invocation by updating the telemetry context."""
+    elif required:
+        raise MeltanoConfigurationError(
+            reason="A Meltano environment must be specified",
+            instruction="Set the `default_environment` option in "
+            "`meltano.yml`, or the `--environment` CLI option",
+        )
 
-    def invoke(self, ctx):
+
+def activate_explicitly_provided_environment(
+    ctx: click.Context, project: Project
+) -> None:
+    """Activate the selected environment if it has been explicitly set.
+
+    Some commands (e.g. `config`, `job`, etc.) do not respect the configured
+    `default_environment`, and will only run with an environment active if it
+    has been explicitly set (e.g. with the `--environment` CLI option).
+
+    Args:
+        ctx: The Click context, used to determine the selected environment.
+        project: The project for which the environment will be activated.
+    """
+    if ctx.obj["is_default_environment"]:
+        logger.info(
+            f"The default environment {ctx.obj['selected_environment']!r} will "
+            f"be ignored for `meltano {ctx.command.name}`. To configure a specific "
+            "environment, please use the option `--environment=<environment name>`."
+        )
+        project.deactivate_environment()
+    else:
+        activate_environment(ctx, project)
+
+
+class CliEnvironmentAction(Enum):
+    """Enum of the different ways in which a Meltano environment can be activated."""
+
+    activate = auto()
+    activate_required = auto()
+    activate_explicitly_provided = auto()
```

Comment:
Per our conversation, some naming ideas. Longer, but doesn't require special training and speaks for itself.

```
activate_required            > environment_required
# use default_environment or explicit environment. otherwise, fail. (meltano run)

activate                     > environment_optional_use_default
# use default_environment or explicit environment. otherwise, no environment

activate_explicitly_provided > environment_optional_ignore_default
# ignore default_enviornment. only active environment if explicitly requested. otherwise, no environment.
```

cc @cjohnhanson

Reply:
Done in 7b779c4.

I also renamed `CliEnvironmentAction` to `CliEnvironmentBehavior`, along with other related `action` -> `behavior` changes, to reflect the updated names.